plugins/modules/ovirt_disk.py

```diff
@@ -420,118 +429,173 @@ def create_transfer_connection(module, transfer, context, connect_timeout=10, re
+            # the database, we can assume that the disk status is already
+            # updated, so we can check it only once.
+            disk_service = (connection.system_service()
+                            .disks_service()
+                            .disk_service(module.params['id']))
+            try:
+                disk = disk_service.get()
+            except sdk.NotFoundError:
+                # Disk verification failed and the system removed the disk.
+                raise RuntimeError(
+                    "Transfer {0} failed: disk {1} was removed"
+                    .format(transfer.id, module.params['id']))
+
+            if disk.status == otypes.DiskStatus.OK:
+                break
+
+            raise RuntimeError(
+                "Transfer {0} failed: disk {1} is '{2}'"
+                .format(transfer.id, module.params['id'], disk.status))
+
+        if transfer.phase == otypes.ImageTransferPhase.FINISHED_SUCCESS:
+            break
+
+        if transfer.phase == otypes.ImageTransferPhase.FINISHED_FAILURE:
+            raise RuntimeError(
+                "Transfer {0} failed, phase: {1}"
+                .format(transfer.id, transfer.phase))
+
+        if time.time() > start + module.params.get('timeout'):
+            raise RuntimeError(
+                "Timed out waiting for transfer {0} to finalize, phase: {1}"
+                .format(transfer.id, transfer.phase))
 
 
 def download_disk_image(connection, module):
-    def _transfer(transfer_service, transfer_connection, transfer_url):
-        BUF_SIZE = 128 * 1024
-        transfer_connection.request('GET', transfer_url.path)
-        r = transfer_connection.getresponse()
-        path = module.params["download_image_path"]
-        image_size = int(r.getheader('Content-Length'))
-        with open(path, "wb") as mydisk:
-            pos = 0
-            while pos < image_size:
-                to_read = min(image_size - pos, BUF_SIZE)
-                chunk = r.read(to_read)
-                if not chunk:
-                    raise RuntimeError("Socket disconnected")
-                mydisk.write(chunk)
-                pos += len(chunk)
-
-    return transfer(
-        connection,
-        module,
-        otypes.ImageTransferDirection.DOWNLOAD,
-        transfer_func=_transfer,
-    )
+    transfers_service = connection.system_service().image_transfers_service()
+    hosts_service = connection.system_service().hosts_service()
+    transfer = start_transfer(connection, module, otypes.ImageTransferDirection.DOWNLOAD)
+    try:
+        extra_args = {}
+        parameters = inspect.signature(client.download).parameters
+        if "proxy_url" in parameters:
+            extra_args["proxy_url"] = transfer.proxy_url
+        if module.params.get('max_workers') and "max_workers" in parameters:
+            extra_args["max_workers"] = module.params.get('max_workers')
+        client.download(
```

Comment:
Slightly inconsistent the opposite order of URL & path in "client.download()" vs. "client.upload()" :(